optuna/pruners/_patient.py

```diff
@@ -0,0 +1,117 @@
+from typing import Optional
+
+import numpy as np
+
+import optuna
+from optuna._experimental import experimental
+from optuna._study_direction import StudyDirection
+from optuna.pruners import BasePruner
+
+
+@experimental("2.8.0")
+class PatientPruner(BasePruner):
+    """Pruner which wraps another pruner with tolerance.
+
+    Example:
+
+        .. testcode::
+
+            import numpy as np
+            from sklearn.datasets import load_iris
+            from sklearn.linear_model import SGDClassifier
+            from sklearn.model_selection import train_test_split
+
+            import optuna
+
+            X, y = load_iris(return_X_y=True)
+            X_train, X_valid, y_train, y_valid = train_test_split(X, y)
+            classes = np.unique(y)
+
+
+            def objective(trial):
+                alpha = trial.suggest_float("alpha", 0.0, 1.0)
+                clf = SGDClassifier(alpha=alpha)
+                n_train_iter = 100
+
+                for step in range(n_train_iter):
+                    clf.partial_fit(X_train, y_train, classes=classes)
+
+                    intermediate_value = clf.score(X_valid, y_valid)
+                    trial.report(intermediate_value, step)
+
+                    if trial.should_prune():
+                        raise optuna.TrialPruned()
+
+                return clf.score(X_valid, y_valid)
+
+
+            study = optuna.create_study(
+                direction="maximize",
+                pruner=optuna.pruners.PatientPruner(optuna.pruners.MedianPruner(), patience=1),
+            )
+            study.optimize(objective, n_trials=20)
+
+    Args:
+        wrapped_pruner:
+            Wrapped pruner to perform pruning when ``PatientPruner`` allows a
+            trial to be pruned. If it is :obj:`None`, this pruner is equivalent to
+            early-stopping taken the intermediate values in the individual trial.
+        patience:
+            Pruning is disabled until the objective doesn't improve for
+            ``patience`` consecutive steps.
+        min_delta:
+            Tolerance value to check whether or not the objective improves.
+            This value should be non-negative for minimization,
+            and it should be non-positive for maximization.
+    """
+
+    def __init__(
+        self, wrapped_pruner: Optional[BasePruner], patience: int, min_delta: float = 0.0
+    ) -> None:
+
+        if patience < 0:
+            raise ValueError("patience cannot be negative but got {}.".format(patience))
+
+        self.wrapped_pruner = wrapped_pruner
```

Comment:
How about documenting this attribute if made public using the `Attributes:` annotation?

Reply:
Indeed, but I think `wrapped_pruned` can be private rather than public because users do not need to access the wrapped pruner.

Reply:
Sounds good too. We can always make it public later on  if needed. Let's prefix the property with an underscore in this PR.